pkg/scanner/local/scan.go

```diff
@@ -73,59 +73,68 @@ func NewScanner(applier Applier, ospkgDetector OspkgDetector) Scanner {
-			log.Logger.Warn(`e.g. files under "/lib/apk/db/", "/var/lib/dpkg/" and "/var/lib/rpm"`)
-		default:
-			return nil, nil, false, xerrors.Errorf("failed to apply layers: %w", err)
-		}
+// Scan scans the artifact and return results.
+func (s Scanner) Scan(target, versionedArtifactID string, versionedBlobIDs []string, options types.ScanOptions) (
+	report.Results, *ftypes.OS, bool, error) {
+	artifactDetail, err := s.applier.ApplyLayers(versionedArtifactID, versionedBlobIDs)
+	switch {
+	case errors.Is(err, analyzer.ErrUnknownOS):
+		log.Logger.Warn("OS is not detected and vulnerabilities in OS packages are not detected.")
+	case errors.Is(err, analyzer.ErrNoPkgsDetected):
+		log.Logger.Warn("No OS package is detected. Make sure you haven't deleted any files that contain information about the installed packages.")
+		log.Logger.Warn(`e.g. files under "/lib/apk/db/", "/var/lib/dpkg/" and "/var/lib/rpm"`)
+	case err != nil:
+		return nil, nil, false, xerrors.Errorf("failed to apply layers: %w", err)
 	}
 
 	var eosl bool
 	var results report.Results
 
-	if utils.StringInSlice("os", options.VulnType) && imageDetail.OS != nil {
-		pkgs := imageDetail.Packages
-		if options.ScanRemovedPackages {
-			pkgs = mergePkgs(pkgs, imageDetail.HistoryPackages)
-		}
-
+	if utils.StringInSlice("os", options.VulnType) && artifactDetail.OS != nil {
 		var result *report.Result
-		result, eosl, err = s.scanOSPkg(target, imageDetail.OS.Family, imageDetail.OS.Name, pkgs)
+		result, eosl, err = s.scanOSPkgs(target, artifactDetail, options)
 		if err != nil {
-			return nil, nil, false, xerrors.Errorf("failed to scan OS packages: %w", err)
-		}
-		if result != nil {
-			if options.ListAllPackages {
-				sort.Slice(pkgs, func(i, j int) bool {
-					return strings.Compare(pkgs[i].Name, pkgs[j].Name) <= 0
-				})
-				result.Packages = pkgs
-			}
+			return nil, nil, false, xerrors.Errorf("unable to scan OS packages: %w", err)
+		} else if result != nil {
 			results = append(results, *result)
 		}
 	}
 
 	if utils.StringInSlice("library", options.VulnType) {
-		libResults, err := s.scanLibrary(imageDetail.Applications, options)
+		libResults, err := s.scanLibrary(artifactDetail.Applications, options)
 		if err != nil {
 			return nil, nil, false, xerrors.Errorf("failed to scan application libraries: %w", err)
 		}
 		results = append(results, libResults...)
 	}
 
-	return results, imageDetail.OS, eosl, nil
+	return results, artifactDetail.OS, eosl, nil
+}
+
+func (s Scanner) scanOSPkgs(target string, detail ftypes.ArtifactDetail, options types.ScanOptions) (
+	*report.Result, bool, error) {
+	pkgs := detail.Packages
+	if options.ScanRemovedPackages {
+		pkgs = mergePkgs(pkgs, detail.HistoryPackages)
+	}
+
+	result, eosl, err := s.detectVulnsInOSPkgs(target, detail.OS.Family, detail.OS.Name, pkgs)
+	if err != nil {
+		return nil, false, xerrors.Errorf("failed to scan OS packages: %w", err)
+	}
+	if result != nil {
```

Comment:
Just my preference.
If result is possible return nil, I prefer to return nil explicitly.